### PR TITLE
Fixes #15233: public key and key-hash comparison not consistent in Rust's version of shared-file API

### PR DESCRIPTION
**Pull request by ArmandSy**
https://issues.rudder.io/issues/15233

### CODE REVIEW

**Comment by ArmandSy** [comment]
PR rebased

File: relay/sources/relayd/src/error.rs [hunk 1]
```diff
@@ -109,6 +110,10 @@ impl Display for Error {
             SetLogLogger(ref err) => write!(f, "Global logger setting error: {}", err),
             InvalidTtl(ref s) => write!(f, "invalid ttl {}", s),
             MissingTargetNodes => write!(f, "missing nodes list"),
+            InvalidHashType => write!(
+                f,
+                "Invalid hash type provided, available hash types : sha1, sha256, sha512"
```
Comment (amousset) [inline relay/sources/relayd/src/error.rs]: you can remove sha1

File: relay/sources/relayd/src/shared_files.rs [hunk 2]
```diff
@@ -119,23 +122,33 @@ pub fn parse_value(key: &str, file: &str) -> Result<String, ()> {
...
-        .collect();
+    pubkey: String,
+    hash_type: String,
+    keyhash: String,
+) -> Result<bool, Error> {
+    let pubkey_pem = get_pubkey(pubkey);
+
+    let rsa_key_from_pem = Rsa::public_key_from_pem_pkcs1(pubkey_pem.as_bytes()).unwrap();
+
+    let public_key_from_pem = PKey::from_rsa(rsa_key_from_pem)?;
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: no need for this step, you should be able to `public_key_to_der` directly on the Rsa

File: relay/sources/relayd/src/shared_files.rs [hunk 3]
```diff
@@ -203,10 +216,10 @@ pub fn parse_hash_from_raw(raw: String) -> String {
         .collect::<String>()
 }
 
-pub fn get_pubkey(metadata: Metadata) -> String {
+pub fn get_pubkey(pubkey: String) -> String {
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: This could return `Result<PKey, opensslerror>`

Comment (amousset) [reply relay/sources/relayd/src/shared_files.rs]: Or Rsa<Public> instead of PKey, it's up to you

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/Cargo.toml [hunk 4]
```diff
@@ -31,7 +31,7 @@ structopt = { version = "0.2", features = [ "paw" ] }
 paw = "1.0"
 toml = "0.5"
 inotify = "0.7"
-openssl = "0.10"
+openssl = "0.10.23"
```
Comment (amousset) [inline relay/sources/relayd/Cargo.toml]: `0.10` is enough

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/src/shared_files.rs [hunk 5]
```diff
@@ -119,23 +158,36 @@ pub fn parse_value(key: &str, file: &str) -> Result<String, ()> {
...
-    let myvec2: Vec<String> = job_config
-        .nodes
-        .read()
-        .expect("Cannot read nodes list")
-        .get_keyhash_from_uuid(&source_uuid)
-        .unwrap()
-        .split(':')
-        .map(|s| s.to_string())
-        .collect();
+    pubkey: String,
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: I'm make this function take a Pkey and a HashType directly

File: relay/sources/relayd/src/shared_files.rs [hunk 6]
```diff
@@ -119,23 +158,36 @@ pub fn parse_value(key: &str, file: &str) -> Result<String, ()> {
...
+    hash_type: String,
+    keyhash: String,
+) -> Result<bool, Error> {
+    let pubkey_pem = get_pubkey(pubkey)?;
 
-    let (_algorithm, keyhash) = (&myvec2[0], &myvec2[1]);
+    let public_key_der = pubkey_pem.public_key_to_der()?;
 
-    hash_value == keyhash
+    let bytes2: &[u8] = &public_key_der;
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: `bytes2` is a strange name, but you can probably avoid it with the `as_bytes()` method.

**Comment by ArmandSy** [comment]
Commit modified

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/src/shared_files.rs [hunk 7]
```diff
@@ -119,23 +157,30 @@ pub fn parse_value(key: &str, file: &str) -> Result<String, ()> {
...
-    let myvec2: Vec<String> = job_config
-        .nodes
-        .read()
-        .expect("Cannot read nodes list")
-        .get_keyhash_from_uuid(&source_uuid)
-        .unwrap()
-        .split(':')
-        .map(|s| s.to_string())
-        .collect();
+    pubkey: openssl::pkey::PKey<Public>,
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: could you `use` it and remove the prefix?

File: relay/sources/relayd/src/shared_files.rs [hunk 8]
```diff
@@ -119,23 +157,30 @@ pub fn parse_value(key: &str, file: &str) -> Result<String, ()> {
...
+    match hash_type {
+        HashType::Sha256 => Ok(HashType::Sha256.hash(public_key_der) == keyhash),
+        HashType::Sha512 => Ok(HashType::Sha512.hash(public_key_der) == keyhash),
+    }
+}
 
-    hash_value == keyhash
+pub fn get_pubkey(
+    pubkey: String,
+) -> Result<openssl::pkey::PKey<Public>, openssl::error::ErrorStack> {
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: same here, it will be more readable

File: relay/sources/relayd/src/shared_files.rs [hunk 9]
```diff
@@ -119,23 +157,30 @@ pub fn parse_value(key: &str, file: &str) -> Result<String, ()> {
...
-        .map(|s| s.to_string())
-        .collect();
+    pubkey: openssl::pkey::PKey<Public>,
+    hash_type: HashType,
+    keyhash: String,
+) -> Result<bool, Error> {
+    let public_key_der: &[u8] = &pubkey.public_key_to_der()?;
 
-    let (_algorithm, keyhash) = (&myvec2[0], &myvec2[1]);
+    match hash_type {
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: no need for this, you can call `hash_type.hash()` directly :)

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/src/shared_files.rs [hunk 10]
```diff
@@ -119,23 +158,30 @@ pub fn parse_value(key: &str, file: &str) -> Result<String, ()> {
...
+        HashType::Sha256 => Ok(HashType::Sha256.hash(public_key_der) == keyhash),
+        HashType::Sha512 => Ok(HashType::Sha512.hash(public_key_der) == keyhash),
+    }
+}
 
-    hash_value == keyhash
+pub fn get_pubkey(
+    pubkey: String,
```
Comment (amousset) [inline relay/sources/relayd/src/shared_files.rs]: nitpick, but it should take `&str` instead of `String` as it does not consume it.

**Comment by ArmandSy** [comment]
Commit modified

**Comment by ArmandSy** [comment]
Commit modified

**Comment by amousset** [comment]
Good! Congratulations for your victory over openssl!

I'll just wait for the test to pass and merge it.

**Pull request merged**